### DISCUSSION TITLE
What is the best way to use pocketbase whith go / templ / htmx

### DISCUSSION THREAD

**Discussion by yoandl** [pocketbase/pocketbase#5307] [Q&A] [answered]
Hi everyone,

First of all, I want to express my gratitude to the author of PocketBase. It’s an amazing project, and I’m really excited to use it. Thank you for creating this fantastic tool!

I’m currently working on a project where I plan to use Go Templ and HTMX for the frontend and PocketBase as the backend. From what I understand, I’ll need to make direct API calls to PocketBase in any case. However, I’m trying to determine the best approach to structure my application. Specifically, I’m wondering if I should use PocketBase as a framework and run a single server, or if it would be better to run two separate servers, one for the backend and one for the frontend.

What do you think about this ?
Thank you!

**Comment by ganigeorgiev** [comment, marked as answer]
Maybe it won't answer directly your question but at the moment PocketBase doesn't have the necessary ergonomics to build SSR apps (at least not out of the box) and I don't recommend using it with htmx or JS SSR frameworks (_for the latter you can read some rants in https://github.com/pocketbase/pocketbase/discussions/3792#discussioncomment-7620032, https://github.com/pocketbase/pocketbase/discussions/4412#discussioncomment-8575834, https://github.com/pocketbase/pocketbase/discussions/5190#discussioncomment-9999444_).

My personal preference is to interact with PocketBase directly from the client-side (e.g. SPA, mobile apps, cli, etc.).
It was designed with this exact use case in mind (_see https://github.com/presentator/presentator/issues/183_). That is also why we have builtin [API rules](https://pocketbase.io/docs/api-rules-and-filters/) and [stateless JSON REST-ish APIs](https://pocketbase.io/docs/api-records/) in the first place. If you are not going to make use of them, then PocketBase may not be the most suitable option for your use case.

Helper builtin methods (for example loading the auth context from a cookie, builtin form submit endpoints for the various auth actions, etc.) could be added eventually in the future but for now remains a low priority and users will have to create them on their own based on their specific use case (_keep in mind that the router interface will change slightly with the next release as part of the [ongoing refactoring](https://github.com/pocketbase/pocketbase/discussions/4355)_).

**Reply by benallfree** [reply]
Could we write JSVM middleware that converts an incoming cookie to the JWT expected by PocketBase? (Sorry I couldn't find where in the PocketBase code it handles the `Authorization` header.)

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand what do you mean.
If you are asking for the code that loads the auth state in the request you could check
https://github.com/pocketbase/pocketbase/blob/ad92992324c3fd0f09c36300890620a0224c6c06/apis/middlewares.go#L212-L254

For v0.23.0-rc the equivalent code could be found in https://github.com/pocketbase/pocketbase/blob/8271452430603cf3051a1cd89bcde01a0fa8fc89/apis/middlewares.go#L192-L220

**Reply by benallfree** [reply]
Yes perfect. Based on the code above, I think it might be possible to write a JSVM middleware handler that loads the auth token from a cookie rather than the `Authorization` header.

If I can get that working, then the PocketPages SSR tool might be able to handle authorization.

**Comment by Sleepful** [comment]
I made an app that implements PB with GoTempl and HTMX.
Gani is absolutely correct that the ergonomics are not there, despite being feasible
You may look at my project as an example:

https://github.com/Sleepful/mymail/tree/main/app/router

This might be a useful example of the problems that are faced when taking this approach.

cc: @ganigeorgiev (no need to reply, I know PB warns against this approach)
